src/graphics/graphics-device.js

```diff
@@ -9,10 +23,288 @@ import { EventHandler } from '../core/event-handler.js';
  * @augments EventHandler
  */
 class GraphicsDevice extends EventHandler {
+    /**
+     * The canvas DOM element that provides the underlying WebGL context used by the graphics device.
+     *
+     * @type {HTMLCanvasElement}
+     */
+    canvas;
```

Comment:
I'm wondering how this code gets executed (and also converted to es5).

This line is declaring a member and could also be initialising it, like `canvas = null;`. But the constructor does `this.canvas = canvas;`. So does the code on this line run after the constructor's call to `super();`, but before the rest of the constructor code?

Reply:
correct, when you step over this in the debugger, those things execute first (and will be set to undefined), before constructor.

Reply:
this is what it looks like in ES5
<img width="596" alt="Screenshot 2022-02-11 at 09 48 45" src="https://user-images.githubusercontent.com/59932779/153570173-2d01493a-33fd-404d-a1d4-e53d4678b6da.png">

Reply:
Interesting.

So could you do `canvas = canvas;` here, instead of doing it in the constructor?

In a class as large as this it probably makes little difference, but in general we don't want to be initialising members twice.

Reply:
So I guess in general we want to declare (and optionally initialise) variables here or in the constructor, but not both?

Reply:
I doubt that would make any perf difference .. but perhaps it'd be useful to test it in a class such as Vec3?

Reply:
I really don't like the idea instance members are initialised twice in our constructors, it's gross.

However reading [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#field_declarations), they say private versions of these (which we hope to use in future) must be declared this way:

```
Note: Private fields can only be declared up-front in a field declaration.
```

So it seems we ultimately won't have a choice and hopefully compiler will optimise away the redundant set.